manager.go

```diff
@@ -53,99 +49,67 @@ type resolverResponse struct {
 // a slight buffer and a slight delay is added to give preference to reusing workers over creating new.
 //
-// A spike in traffic can lead to an increase in the number of workers many of which are subsequently idle. To
-// allow slowly dropping back down to a reasonable pool workers will exit after processing a set number of requests.
-// The original set of workers will always recreate a new worker on exit to avoid the number dropping too much.
+// A small set of workers are long lived to allow some processing to happen at all times and enable a buffered channel
+// Most other workers are not long lived but will only exit when the request channels are empty.
 type resolveManager struct {
 	completeRequests chan completeRequest
-	once             sync.Once
-	quit             chan bool
 	resolveRequests  chan resolveRequest
 }
 
-func newResolveManager(startingWorkers int) *resolveManager {
+func newResolveManager() *resolveManager {
 	manager := &resolveManager{
 		completeRequests: make(chan completeRequest, requestQueueBuffer),
-		once:             sync.Once{},
-		quit:             make(chan bool),
 		resolveRequests:  make(chan resolveRequest, requestQueueBuffer),
 	}
 
-	for i := 0; i < startingWorkers; i++ {
-		go manager.newWorker(true)
+	for i := 0; i < requestQueueBuffer; i++ {
+		go manager.infiniteWorker()
 	}
-
 	return manager
 }
 
 func (manager *resolveManager) completeRequest(req completeRequest) {
-	t := time.NewTimer(workerStartDelay)
-	defer func() {
-		if !t.Stop() {
-			<-t.C
-		}
-	}()
-	for {
-		select {
-		case manager.completeRequests <- req:
-			return
-		case <-t.C:
-			go manager.newWorker(false)
-			t.Reset(workerStartDelay)
-		}
+	select {
+	case manager.completeRequests <- req:
+		return
+	default:
+		go manager.newWorker()
+		manager.completeRequests <- req
 	}
 }
 
-func (manager *resolveManager) resolveRequest(name string, response chan<- resolverResponse, fn FieldResolveFn, params ResolveParams) {
-	req := resolveRequest{
-		fn:       fn,
-		name:     name,
-		params:   params,
-		response: response,
-	}
-
-	t := time.NewTimer(workerStartDelay)
-	defer func() {
-		if !t.Stop() {
-			<-t.C
-		}
-	}()
+func (manager *resolveManager) infiniteWorker() {
 	for {
 		select {
-		case manager.resolveRequests <- req:
-			return
-		case <-t.C:
-			go manager.newWorker(false)
-			t.Reset(workerStartDelay)
+		case req := <-manager.completeRequests:
+			result := completeValueCatchingError(req.eCtx, req.returnType, req.fieldASTs, req.info, req.value)
+			req.response <- completeResponse{index: req.index, result: result}
+		case req := <-manager.resolveRequests:
+			manager.resolve(req)
 		}
 	}
 }
 
-func (manager *resolveManager) newWorker(revive bool) {
-	// This is a simplistic means of dropping the number of workers back down eventually after a spike but other
-	// methods such as added a timer caused a performance impact.
-	for i := 0; i < workerMaxRequests; i++ {
+func (manager *resolveManager) newWorker() {
+	for {
```

Comment:
What is the use for the "for" here?

Reply:
The for keeps this worker running repeatedly until there are no more requests in the channels. Then it will hit the default case and exit.